src/lib.rs

```diff
@@ -1,73 +1,48 @@
 use anyhow::anyhow;
-use base16ct;
-use clap::Parser;
 use fxhash::FxHashMap;
 use md5::{Digest, Md5};
 use std::ffi::{OsStr, OsString};
 
 use regex::Regex;
-use std::fs;
 use std::fs::File;
 use std::io::{self, BufRead};
 use std::path::Path;
 use std::path::PathBuf;
 use tinytemplate::TinyTemplate;
-
 use indicatif::{MultiProgress, ProgressBar, ProgressStyle};
 use std::time::Instant;
 
 
 use crate::types::*;
 use crate::templates::*;
-pub mod templates;
-pub mod types;
-
-#[derive(Parser)]
-#[command(author, version, about, long_about = None)]
-#[command(propagate_version = true)]
-pub struct Cli {
-    path: PathBuf,
-    /// Output directory, defaults to `tl_out`
-    #[arg(short, default_value = "tl_out")]
-    out: PathBuf,
-    /// Delete out directory if it already exists
-    #[arg(long)]
-    overwrite: bool,
-    /// Return non-zero exit code if unrecognized log lines are found.  Mostly useful for unit
-    /// testing.
-    #[arg(long)]
-    strict: bool,
-    /// Don't open browser at the end
-    #[arg(long)]
-    no_browser: bool,
-}
 
+mod templates;
+mod types;
+
+
+pub type ParseOutput = Vec<(PathBuf, String)>;
+
+pub struct ParseConfig {
+    pub strict: bool,
+}
 
-fn main() -> anyhow::Result<()> {
-    let cli = Cli::parse();
-    let path = cli.path;
-    let out_path = cli.out;
 
-    if out_path.exists() {
-        if !cli.overwrite {
-            panic!(
-                "{} already exists, pass --overwrite to overwrite",
-                out_path.display()
-            );
-        }
-        fs::remove_dir_all(&out_path)?;
-    }
-    fs::create_dir(&out_path)?;
 
+pub fn parse_path(path: &PathBuf, config: ParseConfig) -> anyhow::Result<ParseOutput> {
+    let strict = config.strict;
     let file = File::open(path)?;
     let metadata = file.metadata()?;
     let file_size = metadata.len();
+
+    // TODO: abstract out this spinner to not be part of the library
+    // Instead, add a callback trait for CLIs to implement
```

Comment:
@aorenste what do you think the best way to handle this is? The spinner gets updated on every loop of the log parse, but I don't want this in the library. It's also called on various errors with multi.suspend(). 

Maybe I could implement a trait like StatLogger and put the spinner in the callback in the CLI:

```

trait StatLogger {
    fn init(&self) -> ();
    fn log_stats(&mut self, bytes_read: u64, stat: &Stats) -> ();
    // how to do **args in rust?
    fn error(&self, e : Exception, **args) -> (); 
}

pub fn parse_path_with_logger(path: &PathBuf, config: ParseConfig, logger: impl StatLogger) ...

// uses default no-op logger
pub fn parse_path(path: &PathBuf, config: ParseConfig) ...
```
And then call it later in the code when updating and exiting? Or is there a more rust-like way to do it?

Reply:
Yes - I think a trait is the right approach.

I would make sure to either make a trivial null implementation that callers can instantiate or have the call take Option<> so that callers don't have to build their own if they don't care.

Also you should take `&mut impl StatLogger` so it's borrowed instead of owned.

Reply:
Oh - I missed that you had two different APIs for it (one with and one without).  I would just have one API that took `Option<&mut impl StatLogger>`.

Reply:
I'll do this in a separate PR: there's some complexities with the fact that you can't actually condition statically on `impl StatLogger` and also the spinner's side effects need to happen in the function initially. Going to land this change first